src/unix/core.c

```diff
@@ -1517,3 +1526,68 @@ void uv_sleep(unsigned int msec) {
 
   assert(rc == 0);
 }
+
+
+void uv__metrics_update_idle_time(uv_loop_t* loop) {
+  uv__metrics_loop_t* mloop;
+  uint64_t entry_time;
+  uint64_t idle_time;
+  uint64_t exit_time;
+
+  if (!(loop->flags & UV_LOOP_IDLE_TIME))
+    return;
+
+  mloop = uv__metrics_get_mloop(loop);
+
+  /* The thread running uv__metrics_update_idle_time() is always the same
+   * thread that sets provider_entry_time. So it's unnecessary to perform the
+   * atomic operation retrieving this value.
+   * Reason for this check is in case this function has been called a second
+   * time. Since it's always called just after uv__io_poll() to make sure the
+   * idle time has been correctly calculated.
+   */
+  if (mloop->provider_entry_time == 0)
+    return;
+
+  /* Grab the exit_time before starting the atomic operations. */
+  exit_time = uv_hrtime();
+
+  /* It's important to run the operations in this order. It will prevent
+   * accidentally reporting incorrect values and removes the need for a lock.
+   */
+  UV_ATOMIC_EXCHANGE(&mloop->provider_entry_time, &entry_time, 0);
+  UV_ATOMIC_ADD_FETCH(&mloop->idle_time, &idle_time, exit_time - entry_time);
+}
+
+
+void uv__metrics_set_provider_entry_time(uv_loop_t* loop) {
+  uv__metrics_loop_t* mloop;
+  uint64_t now;
+
+  if (!(loop->flags & UV_LOOP_IDLE_TIME))
+    return;
```

Comment:
Maybe asserting here as it should not be called if `UV_LOOP_IDLE_TIME` is not set.

Reply:
But it is always called. To simplify the logic at each call site I placed the check in the function. Otherwise all 14 call locations would need to be wrapped in this same conditional.